pkg/boot/fit/vfit.go

```diff
@@ -0,0 +1,211 @@
+//    P: value
+//    P: algo          (ex. 'sha256,rsa4096', 'pgp')
+//    P: signer-name   (Optional)
+//    P: key-name-hint (Optional)
+//
+// vfit aims to be TOCTTOU-safe by reading files into memory before verifying.
+package fit
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/vfile"
+	"golang.org/x/crypto/openpgp"
+)
+
+var (
+	algs = map[string]crypto.Hash{
+		"MD4":       crypto.MD4,
+		"MD5":       crypto.MD5,
+		"SHA1":      crypto.SHA1,
+		"SHA224":    crypto.SHA224,
+		"SHA256":    crypto.SHA256,
+		"SHA384":    crypto.SHA384,
+		"SHA512":    crypto.SHA512,
+		"RIPEMD160": crypto.RIPEMD160,
+		"SHA3_224":  crypto.SHA3_224,
+		"SHA3_256":  crypto.SHA3_256,
+		"SHA3_384":  crypto.SHA3_384,
+		"SHA3_512":  crypto.SHA3_512,
+	}
+)
+
+type Signature interface {
+	fmt.Stringer
+	// Warning: If the signature does not exist or does not match the keyring,
+	// both the file and a signature error will be returned.
+	Verify([]byte, openpgp.KeyRing) (*bytes.Reader, error)
+}
+
+type PGPSignature struct {
+	name  string // Name of signature Node
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+type RSASignature struct {
+	name  string // Name of signature Node
+	hash  crypto.Hash
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+func (s PGPSignature) String() string {
+	return fmt.Sprintf("PGP Signature - name: %s, signer: '%s', hint: '%s'", s.name, s.signer, s.hint)
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
+func (s PGPSignature) Verify(b []byte, ring openpgp.KeyRing) (*bytes.Reader, error) {
+	r := bytes.NewReader(b)
+	if signer, err := openpgp.CheckDetachedSignature(ring, bytes.NewReader(b), bytes.NewReader(s.value)); err != nil {
+		return r, vfile.ErrUnsigned{Path: s.name, Err: err}
+	} else if signer == nil {
+		return r, vfile.ErrUnsigned{Path: s.name, Err: vfile.ErrWrongSigner{ring}}
+	}
+	return r, nil
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
+func (s RSASignature) Verify(b []byte, ring openpgp.KeyRing) (*bytes.Reader, error) {
+	r := bytes.NewReader(b)
+	keys, err := vfile.GetRSAKeysFromRing(ring)
+	if err != nil {
+		return r, err
+	}
+
+	hashed, err := vfile.CalculateHash(bytes.NewReader(b), s.hash.New())
+	for _, key := range keys {
+		if err = rsa.VerifyPKCS1v15(key, s.hash, hashed, s.value); err == nil {
+			return r, nil
+		}
+	}
+	return r, vfile.ErrUnsigned{Err: vfile.ErrWrongSigner{ring}}
+}
+
+func (s RSASignature) String() string {
+	return fmt.Sprintf("RSA Signature - name: %s, signer: '%s', hint: '%s', hash: '%s'", s.name, s.signer, s.hint, s.hash)
+}
+
+// parseHash cleans and maps a string into a crytpo.Hash.
+// Expected format: "sha256,rsa4096" or "sha1"
```

Comment:
It's unclear to me if this string is supposed to be an agglomeration of all the parameters in some (or no particular) order based on the documentation of this API. Is it an error to pass "rsa4096,sha256"?

Reply:
For hash based algos, it is currently only looking at the first arg for a hash. This seems fine to me and matches the internal format.